### ISSUE TITLE
Ability to wrap task execution loop in the worker threads (worker thread decorator/middleware)

### ISSUE THREAD

**Issue by danielytics** [taskflow/taskflow#404]
**Please describe your feature request.**
I'd like to be able to inject custom logic into the worker threads to be run "around" the execution loop. By around, I mean that I'd like to be able to run code that calls  the loop, rather than just calling a function before and after the loop. Like this:

```cpp
void decorator (Func func) {
   // custom logic before
    func();
   // custom logic after
}
void worker_thread () {
    decorator([](){
        worker_task_loop();
    });
}
```

More concretely, it could be implemented something like this (in the `_spawn` function):

```cpp
    _threads.emplace_back([this] (
      Worker& w, std::mutex& mutex, std::condition_variable& cond, size_t& n
    ) -> void {

      // enables the mapping
      {
        std::scoped_lock lock(mutex);
        _wids[std::this_thread::get_id()] = w._id;
        if(n++; n == num_workers()) {
          cond.notify_one();
        }
      }

      //this_worker().worker = &w;

     user_decorator([this](){  // <------------------------------------- CHANGE MADE HERE
          Node* t = nullptr;
          // must use 1 as condition instead of !done
          while(1) {
    
            // execute the tasks.
            _exploit_task(w, t);
    
            // wait for tasks
            if(_wait_for_task(w, t) == false) {
              return; // <---------------------------------------------- CHANGE MADE HERE
            }
          }
      }); // <---------------------------------------------------------- CHANGE MADE HERE

    }, std::ref(_workers[id]), std::ref(mutex), std::ref(cond), std::ref(n));
```

I'm unsure how `user_decorator` could be injected without changing the existing interface. I suppose it could simply be a function that is passed into the executor constructor, which defaults to a function that does nothing but call its parameter, something like:

```cpp
void default_decorator (std::function<void()> taskLoopFunc) {
    taskLoopFunc();
}
```

To avoid using `std::function`, `this` could instead be passed in as an argument, so the lambda won't need to capture `this` and could decay to a free function:

```cpp
using TaskLoopFunc = void(*)(Executor*);
void default_decorator (TaskLoopFunc taskLoopFunc, Executor* self) {
    taskLoopFunc(self);
}
```

The compiler should inline it, so the end result should be the same as it currently is, unless you replace it with your own logic.

Something like that, anyway. Perhaps there's a better design to achieve the same result.

There are a few possible use cases for this, but for me, I primarily have two cases in mind, which could be implemented like this:

```cpp
void my_decorator (TaskLoopFunc taskLoopFunc, , Executor* self) {
    set_thread_name("worker"); // Name the thread
    try {
        taskLoopFunc(self);
    } catch (...) {} // catch exceptions
}
```
Basically, it would allow you to run custom initialization or termination code (in my case, I want to name the threads, but it could also be used to set thread affinity or other things) and the fact that I can call the loop function (rather than just injecting before and after functions) means I can also wrap the loop in a try/catch block to catch exceptions thrown by tasks for graceful shutdown. You could also rerun the loop after handling the error or whatever else you want instead.

**Comment by tsung-wei-huang** [comment]
Hi @danielytics thank you for the issue. I believe this is doable by extending the constructor of `tf::Executor` to include custom set-up and tear-down methods before each worker thread enters the loop:

```
tf::Executor(size_t N, T&& set_up_thread, U&& tear_down_thread);
```

Extend _spawn to take the two methods, set_up_thread and tear_down_thread
```
set_up_thread();

while(1) {
  // execute the tasks.
  _exploit_task(w, t);
  // wait for tasks
  if(_wait_for_task(w, t) == false) {
    return; // <---------------------------------------------- CHANGE MADE HERE
  }
}

tear_down_thread();
```

What do you think?

**Comment by danielytics** [comment]
Thanks for the quick response and for considering the feature!

That covers the use case of setup and tear down, but it won’t allow me to wrap the loop in a try/catch block to catch exceptions.

If you don’t mind adding such a block to the existing code, then there is an alternative:

```cpp
set_up_thread();

std::exception_ptr exception_ptr = nullptr;
try {
    while(1) {
      // execute the tasks.
      _exploit_task(w, t);
      // wait for tasks
       if(_wait_for_task(w, t) == false) {
         break;  // replace return with break to call tear_down_thread 
       }
    }
} catch (…) {
    exception_ptr =  #std::current_exception();
}

tear_down_thread(exception_ptr);
```

This isn’t quite equivalent to the wrapper function version I described since there’s no way to restart the loop, should you wish to do so. This could of course be modified to provide that but I think that adds a lot of complexity. I personally don’t need the ability to restart, so it’s not a big deal, I just wanted to point out the limitation.

**Comment by tsung-wei-huang** [comment]
@danielytics Thanks for the suggestion. Yes, I think that is a good idea. I am proposing the following abstract class that allows users to customize the thread behavior:

```
class WorkerTraits {
  void set_up();
  void tear_down(std::exception_ptr);
  // ... other potential methods to extend in the future
};

tf::Executor executor(4, WorkerTraits{});
```

Any other thoughts?

**Comment by danielytics** [comment]
No other thoughts, that looks good.

**Comment by tsung-wei-huang** [comment]
This issue has just been addressed in the present master branch and will be part of the future 3.5.0 release. Please also refer to [Change the Worker Property](https://taskflow.github.io/taskflow/ExecuteTaskflow.html#ChangeWorkerProperty).

Feel free to reopen it if any questions or problems. Thank you for your contribution.
